File: static/js/room.js
function getRoomDescription(roomId) {
    fetch('/get_room_description/' + roomId, { method: 'POST' })
        .then(response => response.json())
        .then(data => {
            var description = data.description;
            var interactiveItemsData = data.interactive_items;
            var npcsData = data.npcs;
            var droppedItems = data.dropped_items;

            for (var itemId in interactiveItemsData) {
                var itemHtml = createInteractiveItemHtml(itemId, interactiveItemsData[itemId]);
                description = description.replace("[" + itemId + "]", itemHtml);
            }

            for (var npcId in npcsData) {
                var npcHtml = createNpcHtml(npcId, npcsData[npcId]);
                description = description.replace("[" + npcId + "]", npcHtml);
            }

            var resourceTags = description.match(/\[resource: .+?]/g) || [];
            for (var i = 0; i < resourceTags.length; i++) {
                var tag = resourceTags[i];
                var components = tag.match(/resource: (.+?) \| (\d+) \| (.+?) \| (.+?)]/);
                var resourceHtml = createResourceHtml(tag, components);
                description = description.replace(tag, resourceHtml);
            }

            var droppedItemsHtml = '';
            for (var droppedItemId in droppedItems) {
                var droppedItemData = droppedItems[droppedItemId];
                if (droppedItemData) {
                    droppedItemsHtml += createDroppedItemHtml(droppedItemId, droppedItemData);
                }
            }

            document.getElementById('dropped-items').innerHTML = droppedItemsHtml ? "On the floor, you dropped: " + droppedItemsHtml : '';
            document.getElementById('room-description').innerHTML = description;

            document.querySelectorAll('.interactive-item').forEach(item => {
                item.addEventListener('click', () => inspectItem(item.dataset.itemId));
            });

            document.querySelectorAll('.resource-item').forEach(resource => {
                resource.addEventListener('click', () => pickUpResources(resource.dataset.resourceId, resource.dataset.quantity, roomId, resource.dataset.state));
            });

               document.querySelectorAll('[data-npc-id]').forEach(npcElement => {
                npcElement.addEventListener('click', function() {
                    var npcId = npcElement.getAttribute('data-npc-id');
                    fetch('/dialogues/start_npc/' + roomId + '/' + npcId)
                        .then(response => response.json())
                        .then(data => {
                            displayDialogue(data.dialogue_id, data.dialogue, npcId);
                        });
            });
    });
        });

        
}

window.onload = function() {
    var roomId = document.querySelector('div[data-room-id]').dataset.roomId;
    getRoomDescription(roomId);
    fetchAndUpdateResources();
    fetchAndUpdateInventory();
}

File: engine/rooms/room.py
class Room:
    def __init__(self, name, room_id, base_description, directions, conditions, dynamic_text, state, interactive_items, npcs):
        self.name = name
        self.id = room_id
        self.base_description = base_description
        self.directions = directions
        self.conditions = conditions or {}
        self.dynamic_text = dynamic_text or {}
        self.state = state or {}
        self.interactive_items = interactive_items
        self.npcs = npcs
        self.dropped_items = {}
        
    def get_description(self):
        description = self.base_description
        for key, dynamic_text in self.dynamic_text.items():
            placeholder = '{' + key + '}'
            text = dynamic_text['default']
            for condition, condition_text in dynamic_text['conditions'].items():
                if self.state.get(condition) == 'true':
                    text = condition_text
                    break
            description = description.replace(placeholder, text)
        return description

    def get_parsed_description(self):
        parsed_description = self.get_description()
        interactive_items_data = {}
        for item_id, item in self.interactive_items.items():
            interactive_items_data[item_id] = {
                'name': item.name,
                'description': item.description
            }

        npcs_data = {}
        for npc_id, npc in self.npcs.items():
            npcs_data[npc_id] = {
                'name': npc.name
            }
        
        dropped_items_data = {}
        for item_id, item in self.dropped_items.items():
            dropped_items_data[item_id] = {
                'name': item.name,
            }

        return parsed_description, interactive_items_data, npcs_data, dropped_items_data
        
    def get_interactive_items(self):
        return self.interactive_items

    def set_description(self, description):
        self.description = description

    def get_dropped_items(self):
        return self.dropped_items